### ISSUE TITLE
How to access all annotations of a CtClass including annotations on the super classes?

### ISSUE THREAD

**Issue by vmassol** [INRIA/spoon#3294]
Hi,

I'm trying to get all annotations of a CtClass including annotations on the super classes. Is there a simple way to do that.

Right now I'm struggling with the fact that ctClass.getSuperClass() returns a ctTypeReference (and not a CtClass) and there's no annotation on the ctTypeReference object (even though I do have annotation on the superclass).

Any idea? 

Thanks!

**Comment by vmassol** [comment]
Right now my only idea is to use the current thread CL to load the Class and then call the Java reflection API to get the information but that seems a pity since it should  be the role of Spoon to provide this info.

**Comment by stone-wall** [comment]
From the CtTypeReference, you can call getType() which will get the CtClass which I think should have the annotations on it

**Comment by vmassol** [comment]
@stone-wall Thanks. I've tried to look for a `getType()` method but cannot find on that would return a `CtClass`, see screenshot.

<img width="872" alt="Screenshot 2020-03-07 at 17 06 07" src="https://user-images.githubusercontent.com/628267/76146785-07c60400-6096-11ea-9c6f-a3052e4168d5.png">

Thanks

**Comment by vmassol** [comment]
ok found it. I need to do a cast:

`current = (CtClass) current.getSuperclass().getTypeDeclaration();`

Testing with this, thanks!

**Comment by vmassol** [comment]
Cool it's working, thanks a lot.

So I'd like to repurpose this issue into a request to provide some utility or API in Spoon to make it easier to get all annotations including those from superclasses, for a given `CtClass`. Seems it would be useful and make it easier to use the API.

Right now users have to code this which is a bit cumbersome (TBH I was expecting Spoon to provide it):

```
    private List<CtAnnotation<? extends Annotation>> getAnnotationsIncludingFromSuperclasses(
        CtClass ctClass)
    {
        List<CtAnnotation<? extends Annotation>> annotations = new ArrayList<>(ctClass.getAnnotations());
        CtClass current = ctClass;
        while (current.getSuperclass() != null) {
            current = (CtClass) current.getSuperclass().getTypeDeclaration();
            annotations.addAll(current.getAnnotations());
        }
        return annotations;
    }
```

WDYT? :)

Thanks

**Comment by MartinWitt** [comment]
Following code would be an example using visitor. You would like to have a method `CtType.getAnnotationsIncludingFromSuperclasses` or a predefined visitor like below?
```java
package spoon;

import java.util.ArrayList;
import java.util.List;

import org.junit.Test;

import spoon.reflect.CtModel;
import spoon.reflect.declaration.CtAnnotation;
import spoon.reflect.declaration.CtClass;
import spoon.reflect.visitor.CtAbstractVisitor;
import spoon.support.compiler.VirtualFile;

/**
 * Issue3294
 */
public class Issue3294 {

	private static VirtualFile lower = new VirtualFile(
			"package spoon.issue3294;@Deprecated	public class Lower extends Upper {public void consume(String value) {}}");
	private static VirtualFile upper = new VirtualFile(
			"package spoon.issue3294; @SuppressWarnings(\"all\") public class Upper { }");

	@Test
	public void issue() {
		Launcher launcher = new Launcher();
		launcher.addInputResource(upper);
		launcher.addInputResource(lower);
		CtModel model = launcher.buildModel();
		model.getAllTypes().forEach(v -> v.accept(new CtAbstractVisitor() {
			@Override
			public <T> void visitCtClass(CtClass<T> ctClass) {
				List<CtAnnotation<?>> annotations = new ArrayList<>();
				annotations.addAll(ctClass.getAnnotations());
				CtClass<?> currentClass = ctClass;
				while (currentClass.getSuperclass() != null) {
					annotations.addAll(currentClass.getAnnotations());
					currentClass = (CtClass<?>) currentClass.getSuperclass().getTypeDeclaration();
				}
			}
		}));
	}
}
```

**Comment by vmassol** [comment]
Thanks @MartinWitt for looking into this! :)

I'm coding a `Processor` (extending `AbstractProcessor`). Thus my entry point is the `process()` method from the processor (i.e. `public void process(CtClass<?> ctClass)` to be exact).

I haven't encountered visitors yet. From an API POV, having a `CtClass#getAnnotations(boolean includeSuperClass)` could be nice, WDYT? (and thus also `CtClass#getAnnotation(..., boolean includeSuperClass)`. Or `CtClass#getAnnotationsIncludingSuperClasses()` but then you'd need to double all the `getAnnotations*()`API. Now you may not want to break backward-compatibility either ;)

Another option would be to offer an `AnnotationUtils#getAnnotationIncludingSuperClasses()` but it makes it less easily discoverable from a usage POV. But it's probably the simplest if you don't want to change the API.

\<off topic>
BTW I'm using the following because I don't know how to ask for a specific Annotation (without having the Class object):

```
        boolean hasComponentAnnotation = false;
        boolean hasInstantiationStrategyAnnotation = false;
        boolean hasSingletonAnnotation = false;
        for (CtAnnotation annotation : getAnnotationsIncludingFromSuperclasses(ctClass)) {
            if (COMPONENT_ANNOTATION.equals(annotation.getAnnotationType().getQualifiedName())) {
                hasComponentAnnotation = true;
            } else if (INSTANTIATION_STRATEGY_ANNOTATION.equals(annotation.getAnnotationType().getQualifiedName())) {
                hasInstantiationStrategyAnnotation = true;
            } else if (SINGLETON_ANNOTATION.equals(annotation.getAnnotationType().getQualifiedName())) {
                hasSingletonAnnotation = true;
            }
        }
```

It would probably be nicer if I could simply call `getAnnotation(<my annotation reference here>)`.

Do you know if I could achieve this?
\</off topic>

Thanks!

**Comment by MartinWitt** [comment]
adding a new method to CtType looks the best way currently . `public List<? extends Annotation> getAnnotationsFromSuperTypes(boolean includeClasses, boolean includeInterfaces)`

answer for OT:
You could "simple" use this
```java
	public boolean hasAnnotation(CtClass<?> ctClass, Class<? extends Annotation> annotation) {
		CtClass<?> currentClass = ctClass;
		do {
			if (Objects.nonNull(currentClass.getAnnotation(annotation))) {
				return true;
			}
			currentClass = (CtClass<?>) currentClass.getSuperclass().getTypeDeclaration();
		}
		while (currentClass.getSuperclass() != null);
		return false;
	}
```

**Comment by vmassol** [comment]
@MartinWitt Ok. BTW do you know why I cannot get a superclass from a `CtClass`? I was expecting  that `CtClass#getSuperClass()` would return a `CtClass` and not a `ctTypeReference`... Otherwise it should be named `getSuperClassReference()`, no?

**Comment by MartinWitt** [comment]
sadly no sry. Maybe @monperrus has some knowledge.

**Comment by stone-wall** [comment]
> @MartinWitt Ok. BTW do you know why I cannot get a superclass from a `CtClass`? I was expecting that `CtClass#getSuperClass()` would return a `CtClass` and not a `ctTypeReference`... Otherwise it should be named `getSuperClassReference()`, no?

I agree with you that the naming convention is a bit confusing, that might be worth looking into changing. 

On another note, the more i think about it, the less i believe that a solution along the lines of #3295  is a good idea. i don't think adding a method that can get annotations from all super classes/interfaces is a good idea. i think you would have to set the upper bound in order for that to be viable because you could be potentially dealing with a very large number of super classes up to even java.lang.Object if someone had that package on their scanned classpath. I think probably the best/easiest way to get annotations from the direct super class is to go from the CtClass::getSuperClass::getDeclaratoin::annotations

**Comment by stone-wall** [comment]
Also it looked like maybe earlier you were asking about getting an annotation by it's name? To do that you can use a NamedElementFilter like: `		new NamedElementFilter<>(CtAnnotationType.class, "NonNull");
`

**Comment by vmassol** [comment]
> Also it looked like maybe earlier you were asking about getting an annotation by it's name? To do that you can use a NamedElementFilter like: new NamedElementFilter<>(CtAnnotationType.class, "NonNull");

Thanks. How do you use that? I don't see any signature of `CtClass.getAnnotation()` that accepts a `Filter` as parameter?

Thx

**Comment by vmassol** [comment]
BTW my initial implementation for getting all annotations was buggy since I needed to have annotations defined in extending classes override annotations defined in superclasses. Here's a working impl.:

```
    private Set<CtAnnotation<? extends Annotation>> getAnnotationsIncludingFromSuperclasses(
        CtClass ctClass)
    {
        Set<CtAnnotation<? extends Annotation>> annotations = new HashSet<>();
        Set<String> annotationNames = new HashSet<>();
        CtClass current = ctClass;
        do {
            for (CtAnnotation ctAnnotation : current.getAnnotations()) {
                String annotationName = ctAnnotation.getType().getSimpleName();
                // Note: When they have the same name, we consider that the annotation in the extending class has
                // priority over the the annotation in the super class.
                if (!annotationNames.contains(annotationName)) {
                    annotations.add(ctAnnotation);
                    annotationNames.add(annotationName);
                }
            }
            current = current.getSuperclass() == null ? null : (CtClass) current.getSuperclass().getTypeDeclaration();
        } while (current != null);
        return annotations;
    }
```

**Comment by vmassol** [comment]
This shows that there can be several uses when retrieving all annotations.

**Comment by santos-samuel** [comment]
> @MartinWitt Ok. BTW do you know why I cannot get a superclass from a `CtClass`? I was expecting that `CtClass#getSuperClass()` would return a `CtClass` and not a `ctTypeReference`... 

I believe you cannot get the CtClass directly because it may not be present in your classpath.
From the CtTypeReference you can then call getTypeDeclaration and check if it is != null.
